src/borrow.rs

```diff
@@ -157,7 +150,60 @@ use crate::dtype::Element;
 use crate::error::{BorrowError, NotContiguousError};
 use crate::npyffi::{self, PyArrayObject, NPY_ARRAY_WRITEABLE};
 
-struct BorrowFlags(UnsafeCell<Option<HashMap<usize, isize>>>);
+#[derive(PartialEq, Eq, Hash)]
+struct BorrowKey {
+    range: Range<usize>,
+    data_ptr: usize,
+    gcd_strides: isize,
+}
+
+impl BorrowKey {
+    fn from_array<T, D>(array: &PyArray<T, D>) -> Self
+    where
+        T: Element,
+        D: Dimension,
+    {
+        let range = data_range(array);
+
+        let data_ptr = array.data() as usize;
+        let gcd_strides = reduce(array.strides().iter().copied(), gcd).unwrap_or(1);
```

Comment:
I wonder if this is unnecessarily inefficient. The smaller strides should divide the larger ones, there their GCD should always be the smallest stride, shouldn't it?

Reply:
>  The smaller strides should divide the larger ones,

This is not true when the step size does not divide the dimension along an axis and thereby we do not need to compute the GCD all of strides. This also allows to construct a false conflicts where all conflicting index combinations would be out of bounds. Added a unit test showing this situation.